src/deviation.rs

```diff
@@ -0,0 +1,378 @@
+use ndarray::{ArrayBase, Data, Dimension, Zip};
+use num_traits::{Signed, ToPrimitive};
+use std::convert::Into;
+use std::ops::AddAssign;
+
+use crate::errors::{MultiInputError, ShapeMismatch};
+
+/// An extension trait for `ArrayBase` providing functions
+/// to compute different deviation measures.
+pub trait DeviationExt<A, S, D>
+where
+    S: Data<Elem = A>,
+    D: Dimension,
+{
+    /// Counts the number of indices at which the elements of the arrays `self`
+    /// and `other` are equal.
+    ///
+    /// The following **errors** may be returned:
+    ///
+    /// * `MultiInputError::EmptyInput` if `self` is empty
+    /// * `ShapeMismatch` if `self` and `other` don't have the same shape
+    fn count_eq(&self, other: &ArrayBase<S, D>) -> Result<usize, MultiInputError>
+    where
+        A: PartialEq;
+
+    /// Counts the number of indices at which the elements of the arrays `self`
+    /// and `other` are not equal.
+    ///
+    /// The following **errors** may be returned:
+    ///
+    /// * `MultiInputError::EmptyInput` if `self` is empty
+    /// * `ShapeMismatch` if `self` and `other` don't have the same shape
+    fn count_neq(&self, other: &ArrayBase<S, D>) -> Result<usize, MultiInputError>
+    where
+        A: PartialEq;
+
+    /// Computes the [squared L2 distance] between `self` and `other`.
+    ///
+    /// ```text
+    ///  n
+    ///  ∑  |aᵢ - bᵢ|²
+    /// i=1
+    /// ```
+    ///
+    /// where `self` is `a` and `other` is `b`.
+    ///
+    /// The following **errors** may be returned:
+    ///
+    /// * `MultiInputError::EmptyInput` if `self` is empty
+    /// * `ShapeMismatch` if `self` and `other` don't have the same shape
+    ///
+    /// [squared L2 distance]: https://en.wikipedia.org/wiki/Euclidean_distance#Squared_Euclidean_distance
+    fn sq_l2_dist(&self, other: &ArrayBase<S, D>) -> Result<A, MultiInputError>
+    where
+        A: AddAssign + Clone + Signed;
+
+    /// Computes the [L2 distance] between `self` and `other`.
+    ///
+    /// ```text
+    ///    n
+    /// √  ∑  |aᵢ - bᵢ|²
```

Comment:
Not a pro when it comes to this kind of stuff, but the square root here is not very visible.

Reply:
Thanks. I pushed a commit that adds brackets. Does this help?

Reply:
It does!